pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,1438 @@
+	clone, err := conversion.NewCloner().DeepCopy(volume)
+	if err != nil {
+		return fmt.Errorf("Error cloning pv: %v", err)
+	}
+	volumeClone, ok := clone.(*api.PersistentVolume)
+	if !ok {
+		return fmt.Errorf("Unexpected volume cast error : %v", volumeClone)
+	}
+
+	if ctrl.hasAnnotation(volume.ObjectMeta, annBoundByController) {
+		// The volume was bound by the controller.
+		volumeClone.Spec.ClaimRef = nil
+		delete(volumeClone.Annotations, annBoundByController)
+		if len(volumeClone.Annotations) == 0 {
+			// No annotations look better than empty annotation map (and it's easier
+			// to test).
+			volumeClone.Annotations = nil
+		}
+	} else {
+		// The volume was pre-bound by user. Clear only the binging UID.
+		volumeClone.Spec.ClaimRef.UID = ""
+	}
+
+	newVol, err := ctrl.kubeClient.Core().PersistentVolumes().Update(volumeClone)
+	if err != nil {
+		glog.V(5).Infof("Updating PersistentVolume[%s]: rollback failed: %v", volume.Name, err)
+		return err
+	}
+	glog.V(5).Infof("Updating PersistentVolume[%s]: rolled back", newVol.Name)
+
+	// Update the status
+	_, err = ctrl.updateVolumePhase(newVol, api.VolumeAvailable)
+	return err
+
+}
+
+func (ctrl *PersistentVolumeController) getClaimStatusForLogging(claim *api.PersistentVolumeClaim) string {
+	everBound := ctrl.hasAnnotation(claim.ObjectMeta, annWasEverBound)
+	boundByController := ctrl.hasAnnotation(claim.ObjectMeta, annBoundByController)
+
+	return fmt.Sprintf("phase: %s, bound to: %q, wasEverBound: %v, boundByController: %v", claim.Status.Phase, claim.Spec.VolumeName, everBound, boundByController)
+}
+
+func (ctrl *PersistentVolumeController) getVolumeStatusForLogging(volume *api.PersistentVolume) string {
+	boundByController := ctrl.hasAnnotation(volume.ObjectMeta, annBoundByController)
+	claimName := ""
+	if volume.Spec.ClaimRef != nil {
+		claimName = fmt.Sprintf("%s/%s (uid: %s)", volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name, volume.Spec.ClaimRef.UID)
+	}
+	return fmt.Sprintf("phase: %s, bound to: %q, boundByController: %v", volume.Status.Phase, claimName, boundByController)
+}
+
+// reclaimVolume implements volume.Spec.PersistentVolumeReclaimPolicy and
+// starts appropriate reclaim action.
+func (ctrl *PersistentVolumeController) reclaimVolume(volume *api.PersistentVolume) error {
+	switch volume.Spec.PersistentVolumeReclaimPolicy {
+	case api.PersistentVolumeReclaimRetain:
+		glog.V(5).Infof("reclaimVolume[%s]: policy is Retain, nothing to do", volume.Name)
+
+	case api.PersistentVolumeReclaimRecycle:
+		glog.V(5).Infof("reclaimVolume[%s]: policy is Recycle", volume.Name)
+		ctrl.scheduleOperation("recycle-"+string(volume.UID), ctrl.recycleVolumeOperation, volume)
+
+	case api.PersistentVolumeReclaimDelete:
+		glog.V(5).Infof("reclaimVolume[%s]: policy is Delete", volume.Name)
+		ctrl.scheduleOperation("delete-"+string(volume.UID), ctrl.deleteVolumeOperation, volume)
+
+	default:
+		// Unknown PersistentVolumeReclaimPolicy
```

Comment:
Note: make sure that we validate this as an enum in the API server, this should always be unreachable code.